### ISSUE TITLE
Zenject Warning: It is bad practice to call Inject/Resolve/Instantiate before all the Installers have completed!

### ISSUE THREAD

**Issue by gregee123** [EcsRx/ecsrx.unity#12]
This happens to be from view-binding-scene example.

Yes, code still works but wouldn't it be nice to not have this warning?

Seems like AutoRegisterSystemsInstaller.RegisterSystems() shouldn't be called from within InstallBindings(). It seems like a cheeky hack to register systems while bindings are installed. Certainly requires for the AutoRegisterSystemsInstaller to be dropped onto SceneContext -> Installers in the last position. Perhaps it's better to just use AutoBindSystemsInstaller and use a separate MonoBehaviour based class to run RegisterSystems() if it's possible to assure it'd be run after all bindings have been installed?

Zenject Warning: It is bad practice to call Inject/Resolve/Instantiate before all the Installers have completed!  This is important to ensure that all bindings have properly been installed in case they are needed when injecting/instantiating/resolving.  Detected when operating on type 'ISystem'
UnityEngine.Debug:LogWarning(Object)
ModestTree.Log:Warn(String, Object[]) (at Assets/Zenject/Source/Internal/Log.cs:47)
Zenject.DiContainer:CheckForInstallWarning(InjectContext) (at Assets/Zenject/Source/Main/DiContainer.cs:466)
Zenject.DiContainer:ResolveAll(InjectContext) (at Assets/Zenject/Source/Main/DiContainer.cs:407)
Zenject.DiContainer:ResolveAll(InjectContext) (at Assets/Zenject/Source/Main/DiContainer.cs:1803)
Zenject.DiContainer:ResolveAll(Object, Boolean) (at Assets/Zenject/Source/Main/DiContainer.cs:1795)
Zenject.DiContainer:ResolveAll(Object) (at Assets/Zenject/Source/Main/DiContainer.cs:1788)
Zenject.DiContainer:ResolveAll() (at Assets/Zenject/Source/Main/DiContainer.cs:1776)
EcsRx.Unity.Installers.AutoRegisterSystemsInstaller:RegisterSystems() (at Assets/EcsRx/Unity/Installers/AutoRegisterSystemsInstaller.cs:28)
EcsRx.Unity.Installers.AutoRegisterSystemsInstaller:InstallBindings() (at Assets/EcsRx/Unity/Installers/AutoRegisterSystemsInstaller.cs:23)
Zenject.DiContainer:InstallInstallerInternal(IInstaller) (at Assets/Zenject/Source/Main/DiContainer.cs:680)
Zenject.DiContainer:Install(IInstaller, IEnumerable`1) (at Assets/Zenject/Source/Main/DiContainer.cs:535)
Zenject.DiContainer:Install(IInstaller) (at Assets/Zenject/Source/Main/DiContainer.cs:516)
Zenject.Context:InstallInstallers(Dictionary`2) (at Assets/Zenject/Source/Install/Contexts/Context.cs:158)
Zenject.Context:InstallInstallers() (at Assets/Zenject/Source/Install/Contexts/Context.cs:117)
Zenject.SceneContext:InstallBindings(InitialComponentsInjecter) (at Assets/Zenject/Source/Install/Contexts/SceneContext.cs:232)
Zenject.SceneContext:RunInternal() (at Assets/Zenject/Source/Install/Contexts/SceneContext.cs:183)
Zenject.SceneContext:Run() (at Assets/Zenject/Source/Install/Contexts/SceneContext.cs:134)
Zenject.SceneContext:Awake() (at Assets/Zenject/Source/Install/Contexts/SceneContext.cs:103)

**Comment by grofit** [comment]
It seems this _may_ be doable by just overriding the `Start` method within MonoInstallers, I think this would then allow the bindings to be registered at install time but the registrations to only be fired once it has installed, I dont know if that has any adverse behaviour elsewhere, it seems to work fine in the example projects, will need to do some more testing on that.

**Comment by grofit** [comment]
Nope the start wont work as a simple fix as this means that it is resolved after other `[Inject]` tagged methods, such as `GameStarted` within the `EcsRxApplicaiton` implementation, so the solution is to either put it back into there and just agree its bad practice but it works, or have everyone do more work on the application side to register all systems their own way.

If you are on tomorrow I am happy to discuss options in gitter

**Comment by gregee123** [comment]
Can't you move RegisterSystems() with List<string> SystemNamespaces to EcsRxApplication and call it from within Init() somewhere? I don't suppose it'd be to late or is it?

**Comment by grofit** [comment]
Here is my response from gitter:

---

@gregee123 currently I have added a method to EcsRxApplication which automatically registers all bound systems
that will at least allow you to use the AutoBindAllSystems installer, but then just do one line to get the other systems to be registered
problem is I didnt want that class to be heavy, so I am trying to find a nicer way to do it

---

The auto registers originally were just meant to be a quick attempt at providing some simple conventional way to just add systems and have them register for the right scenes in bulk. So this way I get to keep the autobinding installer which lets the dev tie the installer to the systems then just bulk register them all in the app. I am still not 100% happy with it but its at least still giving you the benefits of being able to configure it for the most part in the unity side, as the main idea was that people would make their own `MonoInstaller` instances for their custom loading conventions, so for the moment this will at least allow simple use cases to persist with little effort, but still allow people to make custom installers and register afterwards.

I did think about using reflection to just get all the systems in the namespace and then registering them after the bind, but then I realised that without the DI resolving them their dependencies would not be met, so its a bit of a problem there. If you come up with any better ways of solving it I am open to it, but I may just release the above approach for now and then look at doing an improved way going forward, assuming you are happy with my approach?
